### DISCUSSION TITLE
Access random objects in json

### DISCUSSION THREAD

**Discussion by mementoseeds** [nlohmann/json#2820] [Q&A] [answered]
Hi, I want to implement a feature in one of my projects which requires being able to select a specific amount of randomly selected items from a json object.

For example, let's say I have this json document
```
{
    "item1": "value1",
    "item2": "value2",
    "item3": "value3"
}
```
and I want to select 2 random keys from it. Then the selected keys could be "item1" and "item2", or "item1" and "item3", or "item2" and "item3". Is it possible to do that?

Alternatively, is it possible to access json objects through an index? For example with the above document, json[0] could return the key "item1". But if the document were reordered like this
```
{
    "item2": "value2",
    "item1": "value1",
    "item3": "value3"
}
```
then json[0] would return "item2".

One way or another, I wish to be able to select a specific amount of objects from a json document chosen at random - either through a built in function or through a number index where the number is randomly generated. Is any of this possible?

Thanks in advance.

**Comment by nlohmann** [comment, marked as answer]
You can do it like explained in https://www.techiedelight.com/get-random-value-stl-containers-cpp/:

Example:

```cpp
#include <iostream>
#include "json.hpp"

using json = nlohmann::json;

const json& get_random(const json& j)
{
    auto it = j.cbegin();
    int random = rand() % j.size();
    std::advance(it, random);
 
    return *it;
}

int main()
{
    json val = R"(
    {
        "item1": "value1",
        "item2": "value2",
        "item3": "value3"
    }
    )"_json;
    
    std::cout << get_random(val) << std::endl;
    std::cout << get_random(val) << std::endl;
    std::cout << get_random(val) << std::endl;
}
```

**Reply by mementoseeds** [reply]
This method seems to always return the same results. For example,  when I run the program once, I get a certain amount of random items. But the next time I run it I get the same items.

I also tried the other algorithms in that page, but they had the same behavior. Is it possible to introduce more randomness?

EDIT: On second thought, this method works just fine. The problem was that `std::rand()` was not random enough, so for my Qt project I instead used `QRandomGenerator::global()->generate()`.

Thanks for the help!

**Reply by gregmarr** [reply]
https://en.cppreference.com/w/cpp/numeric/random/rand
Sounds like you're not calling `std::srand()` first, as that would result in exactly the behavior you are describing, deterministic pseudo-random numbers.
